src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala

```diff
@@ -28,11 +28,103 @@ import scala.concurrent.{ Await, ExecutionContext, Future }
 import scala.util.control.NonFatal
 import scala.util.{ Failure, Success }
 
+trait AssignDynamicServiceLogic extends StrictLogging {
+
+  val config: GroupManagerConfig
+
+  /**
+    * Checks whether newApp is new or changed.
+    * @param originalApps A map of the original apps form before the update.
+    * @param newApp The new app that is tested.
+    * @return true if app is new or an updated, false otherwise.
+    */
+  def changedOrNew(originalApps: Map[AppDefinition.AppKey, AppDefinition], newApp: AppDefinition): Boolean = {
+    originalApps.get(newApp.id) match {
+      case None => true // new app
+      case Some(oldApp) => oldApp.isUpgrade(newApp)
+    }
+  }
+
+  def assignDynamicServicePorts(from: RootGroup, to: RootGroup): RootGroup = {
+    val portRange = Range(config.localPortMin(), config.localPortMax())
+    var taken = from.transitiveApps.flatMap(_.servicePorts) ++ to.transitiveApps.flatMap(_.servicePorts)
```

Comment:
Since we're trying to get this code fast, what do you think about using a mutable set here and initialize it with a while loop?

Reply:
With the changes below the updates to this map should be done for changed apps/changed ports. I wonder if a mutable structure would make any difference here. But then again - we have a benchmark now so we can easily test - data beats opinion!

Reply:
I think the improvement would be minimal since 30% of the time is spent in the RootGroup creation.